### PR TITLE
BUG: pivot_table strings as aggfunc

### LINKED ISSUE #18713: BUG: pivot_table aggfunc should accept string function-likes
Virtually everywhere we allow strings to substitue for functions in ``.agg``, ``.apply`` and such, straightforward to add here. Just need to call https://github.com/pandas-dev/pandas/blob/master/pandas/core/base.py#L627.

example from ``pd.pivot_table`` doc-string

```
In [11]: >>> df = pd.DataFrame({"A": ["foo", "foo", "foo", "foo", "foo",
    ...: ...                          "bar", "bar", "bar", "bar"],
    ...: ...                    "B": ["one", "one", "one", "two", "two",
    ...: ...                          "one", "one", "two", "two"],
    ...: ...                    "C": ["small", "large", "large", "small",
    ...: ...                          "small", "large", "small", "small",
    ...: ...                          "large"],
    ...: ...                    "D": [1, 2, 2, 3, 3, 4, 5, 6, 7]})
    ...: 

In [12]: df
Out[12]: 
     A    B      C  D
0  foo  one  small  1
1  foo  one  large  2
2  foo  one  large  2
3  foo  two  small  3
4  foo  two  small  3
5  bar  one  large  4
6  bar  one  small  5
7  bar  two  small  6
8  bar  two  large  7
```

works
```
In [13]: pd.pivot_table(df, values='D', index=['A', 'B'],
    ...:                      columns=['C'], aggfunc=[np.sum, np.mean])
 Out[13]: 
          sum        mean      
C       large small large small
A   B                          
bar one   4.0   5.0   4.0   5.0
    two   7.0   6.0   7.0   6.0
foo one   4.0   1.0   2.0   1.0
    two   NaN   6.0   NaN   3.0
```

fails
```
In [14]: pd.pivot_table(df, values='D', index=['A', 'B'],
    ...:                      columns=['C'], aggfunc=['sum', 'mean'])
AttributeError: 'str' object has no attribute '__name__'
```

### PR DESCRIPTION
**Pull request by bobhaffner**
- [x] closes #18713
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.22.0.txt [hunk 1]
```diff
@@ -321,7 +321,8 @@ Reshaping
 - Bug in :func:`DataFrame.stack` which fails trying to sort mixed type levels under Python 3 (:issue:`18310`)
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Bug in :func:`Series.rank` where ``Series`` containing ``NaT`` modifies the ``Series`` inplace (:issue:`18521`)
--
+- Bug in :func:`Dataframe.pivot_table` which fails when you pass a string for aggfunc arg (:issue:`18713`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.22.0.txt]: We generally don't use the second-person (or any such person-pronouns in general).

**Comment by bobhaffner** [comment]
Thanks for the feedback @gfyoung.

File: pandas/tests/reshape/test_pivot.py [hunk 2]
```diff
@@ -1109,6 +1109,20 @@ def test_pivot_margins_name_unicode(self):
...
+                                     aggfunc=func)
+        result = f('sum')
+        expected = f(np.sum)
+        tm.assert_frame_equal(result, expected)
+        result = f(['mean', 'std'])
+        means = f(np.mean)
+        stds = f(np.std)
+        expected = concat([means, stds], keys=['mean', 'std'], axis=1)
+        tm.assert_frame_equal(result, expected)
+
```
Comment (gfyoung) [inline pandas/tests/reshape/test_pivot.py]: I like what you're testing here, but I would modify a couple of things:

1) Add newlines after each `tm.assert_frame_equal`.  It makes it easier to read.
2) Construct the output explicitly using the `DataFrame` constructor (and check against that too).  At this point, I know for example that `f('sum')` and `f(np.sum)` should be the same, but I don't know what the actual result is.  e.g.

~~~python
result = f('sum')
func_result = f(np.sum)
expected = DataFrame(...)
tm.assert_frame_equal(result, func_result)  # consistency
tm.assert_frame_equal(result, expected)  # correctness
~~~

Comment (jreback) [reply pandas/tests/reshape/test_pivot.py]: you can parametrize this test by passing these functions in as well

something like

```
@pytest.mark.parametrize("f, f_numpy", [('sum', np.sum), ('mean', np.mean'), ('std', lambda x: np.std(x, ddof=0))])
def test_pivot_func_string(self, f, f_numpy):
```

note that ``std`` will resolve to pandas ``std`` and so ``np.std`` needs handling here (for the test)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_pivot.py (hunk 2 above)
Comment (bobhaffner) [reply pandas/tests/reshape/test_pivot.py]: Thank you, both,  I'll incorporate these changes

**Comment by bobhaffner** [comment]
`test_pivot_string_as_func` is an attempt at @gfyoung's suggestion regarding comparing to a frame created from `DataFrame()` 

`test_pivot_string_func_vs_func` is an attempt at @jreback's suggestion of using `@pytest.mark.parametrize` to compare string_funcs to numpy funcs

Separating the cases into two functions seemed like the right thing to do.  As opposed to one rather long one with multiple `DataFrame` statements

File: pandas/tests/reshape/test_pivot.py [hunk 3]
```diff
@@ -1109,6 +1109,48 @@ def test_pivot_margins_name_unicode(self):
...
+                              ('sum', 'C', 'one'): {'bar': 15, 'foo': 13},
+                              ('sum', 'C', 'two'): {'bar': 7, 'foo': 20}},
+                             columns=mi).rename_axis('A')
+        tm.assert_frame_equal(result, expected)
+
+    funcs = [('sum', np.sum), ('mean', np.mean), ('std', np.std),
+             (['sum', 'mean'], [np.sum, np.mean]),
+             (['sum', 'std'], [np.sum, np.std]),
+             (['std', 'mean'], [np.std, np.mean])]
+
```
Comment (gfyoung) [inline pandas/tests/reshape/test_pivot.py]: Remove newline.

File: pandas/tests/reshape/test_pivot.py [hunk 4]
```diff
@@ -1109,6 +1109,48 @@ def test_pivot_margins_name_unicode(self):
...
+    @pytest.mark.parametrize("f, f_numpy", funcs)
+    def test_pivot_string_func_vs_func(self, f, f_numpy):
+        # GH #18713
```
Comment (gfyoung) [inline pandas/tests/reshape/test_pivot.py]: Add that this test is for consistency purposes.

File: pandas/tests/reshape/test_pivot.py [hunk 5]
```diff
@@ -1109,6 +1109,48 @@ def test_pivot_margins_name_unicode(self):
         expected = pd.DataFrame(index=index)
         tm.assert_frame_equal(table, expected)
 
+    def test_pivot_string_as_func(self):
+        # GH #18713
```
Comment (gfyoung) [inline pandas/tests/reshape/test_pivot.py]: Add that this test is for correctness.

**Comment by gfyoung** [comment]
> Separating the cases into two functions seemed like the right thing to do. As opposed to one rather long one with multiple DataFrame statements.

Seems reasonable to me.

File: pandas/tests/reshape/test_pivot.py (hunk 3 above)
Comment (bobhaffner) [reply pandas/tests/reshape/test_pivot.py]: This is a weird one, flake8 likes that newline

**Comment by bobhaffner** [comment]
Just following up on this to make sure no other changes are required.  No worries if folks are busy

File: pandas/tests/reshape/test_pivot.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/reshape/test_pivot.py]: instead of using funcs, simply put this in the parametrize

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt [hunk 6]
```diff
@@ -341,7 +341,8 @@ Reshaping
 - Bug in :func:`DataFrame.stack` which fails trying to sort mixed type levels under Python 3 (:issue:`18310`)
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Bug in :func:`Series.rank` where ``Series`` containing ``NaT`` modifies the ``Series`` inplace (:issue:`18521`)
--
+- Bug in :func:`Dataframe.pivot_table` which fails when the ``aggfunc`` arg is of type string.  The behavior is now consistent with other methods like ``agg`` and ``apply`` (:issue:`18713`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: rebase on master. can you move to 0.23 (docs were renamed), prob easiest to just check this file from master and past in new one

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by bobhaffner** [comment]
Very interesting.  Looks like the rebase handled the whatsnew change quite nicely

```
(pandas-dev) Bobs-MacBook-Pro:pandas-dev bob$ git rebase upstream/master
First, rewinding head to replay your work on top of it...
Applying: pivot_table strings as aggfunc
Using index info to reconstruct a base tree...
A	doc/source/whatsnew/v0.22.0.txt
.git/rebase-apply/patch:11: trailing whitespace.
- 
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging doc/source/whatsnew/v0.23.0
Applying: fixed pep8 issues
Applying: fixed whatsnew
Using index info to reconstruct a base tree...
A	doc/source/whatsnew/v0.22.0.txt
Falling back to patching base and 3-way merge...
Auto-merging doc/source/whatsnew/v0.23.0
Applying: updated tests
Applying: added comments to test funcs
Applying: removed list usage in test parametrize
```

**Comment by gfyoung** [comment]
@bobhaffner : This is `git` - you should expect great things from it :smile: (unless you're pushing large files OR operating at big-tech-company-scale... :0 )

**Comment by bobhaffner** [comment]
@gfyoung Haha, gotcha  :-)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by bobhaffner** [comment]
> lgtm. ping on green.

@jreback is this intended for me or some kind of shorthand for a notification process?   At any rate, ping   :-)    

Looks like there's a conflict with the whatsnew piece being changed back to the old name.  I can follow suit or wait till this is sorted out.

**Comment by jreback** [comment]
thanks @bobhaffner 

yes 'ping on green' is a notification of you to ping me when this has passed and all buttons are green.

**Comment by bobhaffner** [comment]
Thanks @jreback and @gfyoung!!!

**Pull request merged**